### ISSUE TITLE
Add `copas.select()` API

### ISSUE THREAD

**Issue by azdle** [keplerproject/copas#92]
I'm curious about your thoughts on adding `copas.select()` call that would function like luasocket's `socket.select()` but allow a copas thread to block on multiple sockets' readyness. This seems like this would be beneficial both for porting code from blocking sockets to copas & generally for people who are used to non-blocking networking with select directly to develop new code on copas.

I wanted to get an opinion early because it seems like copas wasn't really written with something like this in mind. Right now it's only possible to yield a single socket (with a single readyness queue), so it would probably require quite a bit of low-level change. Is that an idea you'd even be willing to entertain? Or is it something you've thought about the best way to accomplish before?

**Comment by Tieske** [comment]
what would the usecase be?

**Comment by azdle** [comment]
The use case I'm planning for right now is in the realm of IoT. We have lots of end devices each with an open socket for sending requests, getting responses, and getting async alerts. Then, we have a separate socket (for each device) that goes "up", on which we send and receive parsed events to the rest of the system. I would like to have a single copas thread for each device that is able to block waiting for async events coming from both directions, acting as a sort of translation proxy.

I know I could currently accomplish this with two threads, one for "up" and one for "down", but that runs into issues when devices expect both request-response & async events on the same socket. It would be much cleaner to have both sockets handled from the same thread, processing whichever is ready first That would allow me to keep both data and logic locally in the thread, simplifying things greatly.

It would also be possible to go hard the other way and introduce more threads & some sort of message passing system, basically pseudo sockets, that would allow me to build a actor system. This would also require changes to copas to allow the use of pseudo sockets and I felt that this was much less in line with the spirit of coaps than implementing another call that already exists in luasocket.

**Comment by Tieske** [comment]
I get the issue (I think...). I've been preparing for a similar case, that's why I added the lock. essentially everything can be done async, except if there is a specific request-response sequence to finish first, that shouldn't be interleaved with async messages.

Can't you wrap individual transactions in a lock?

**Comment by azdle** [comment]
Oh, it's definitely possible. I'm just worried it's much more error prone. Especially since, as far as I'm aware, there's no way to really enforce the use of locking resources with lua. (I mean that in the way that I'm used to mutexes working in rust, where there is no way to have access to a resource without holding the lock that contains it.)

I'm only writing the system, not most of the code that it's going to run, so I want to make it as fool proof as possible. (And even if I was writing it, I'd prefer to use something more fool proof so I have less thinking overhead.) And I think that not having to deal with any thread syncing stuff and just throw everything in a single thread makes it overall easier to reason about behavior.

**Comment by Tieske** [comment]
I'm hesitant because the Copas code is already hard to reason about. Adding a `select` like that is going to be a big change.

> I mean that in the way that I'm used to mutexes working in rust, where there is no way to have access to a resource without holding the lock that contains it

essentially in Lua all code runs atomically and no locks are needed. Except if you use an implicit yielding operation (sleep, read/write, etc)

If you create a class/function that will transparently handle the locks, the users would still have a single/simple way of accessing is safely.

**Comment by azdle** [comment]
> I'm hesitant because the Copas code is already hard to reason about. Adding a `select` like that is going to be a big change.

I totally get that, that's why I opened this issue. For all I knew I'd get "I really want that & here's exactly how I was planning on refactoring things to make that work cleanly." for a response.

> > I mean that in the way that I'm used to mutexes working in rust, where there is no way to have access to a resource without holding the lock that contains it
> 
> essentially in Lua all code runs atomically and no locks are needed. Except if you use an implicit yielding operation (sleep, read/write, etc)

I think I didn't explain what I was trying to say well enough. I basically meant that lock management is easier to mess up because you don't *have* to lock to try to (for instance) write to a socket, you don't *have* to remove all references to the socket to unlock the lock, and (extending a bit beyond what I actually wrote) you don't *have* to unlock the lock to carry on your merry way after you done.

> If you create a class/function that will transparently handle the locks, the users would still have a single/simple way of accessing is safely.

Sure, that would be ideal and trying to do that basically my fall-back plan. But, I'm not sure it's even possible to do such that it works for even the naive case of every different device topology. Whereas just keeping everything in a single thread makes those cases just work.

I'll probably punt on this for a bit & then do some playing around with it later. If I'm able to get something that I think is clean, and unless I find something simpler, I'll come back and get your opinion when I have something more concrete to propose.
